cluster.go

```diff
@@ -1021,6 +1038,52 @@ func (c *Cluster) StateSync(ctx context.Context) error {
 	return nil
 }
 
+// unpinIfExpired returns true if given pin was unpinned.
+func (c *Cluster) unpinIfExpired(ctx context.Context, pin *api.Pin, trustedPeers []peer.ID, timeNow time.Time, hashCache utils.HashCache) bool {
+	if pin.PinOptions.ExpireAt.IsZero() ||
+		pin.PinOptions.ExpireAt.After(timeNow) ||
+		!c.isClosest(ctx, pin.Cid, trustedPeers, hashCache) {
+		return false
+	}
+
+	logger.Infof("StateSync: unpinning %s, pin expired at %s", pin.Cid, pin.PinOptions.ExpireAt.Format(time.RFC3339))
+	_, err := c.Unpin(ctx, pin.Cid)
+	if err != nil {
+		logger.Warning(err)
+		return false
+	}
+
+	return true
+}
+
+func (c *Cluster) updateRemotePins(ctx context.Context, pin *api.Pin, p *api.PinInfo) error {
+	var err error
+	allocatedHere := pin.ReplicationFactorMin == -1 || containsPeer(pin.Allocations, c.id)
+
+	switch {
+	case p.Status == api.TrackerStatusRemote && allocatedHere:
+		logger.Debugf("StateSync: Tracking %s locally (currently remote)", p.Cid)
+		err = c.tracker.Track(ctx, pin)
+	case p.Status == api.TrackerStatusPinned && !allocatedHere:
+		logger.Debugf("StateSync: Tracking %s as remote (currently local)", p.Cid)
+		err = c.tracker.Track(ctx, pin)
+	}
+
+	return err
+}
+
+// isClosest tells if this cluster peer is the closest to given cid.
+func (c *Cluster) isClosest(ctx context.Context, ci cid.Cid, trustedPeers []peer.ID, hashCache utils.HashCache) bool {
+	ciHash := utils.ConvertKey(ci.KeyString())
+	myDistance := utils.XOR(ciHash, hashCache.ConvertPeerID(c.id))
+	for _, p := range trustedPeers {
+		if !myDistance.Less(utils.XOR(ciHash, hashCache.ConvertPeerID(p))) {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
Would it make sense to have a 

```go
type distanceChecker struct {
}

func buildDistanceChecker(local peer.ID, otherPeers []peer.ID) distanceChecker {
}

func (dc distanceChecker) closest(cid.Cid) bool {

}
```

that wraps the whole hashcache/xoring/hashing functionality inside? I find this code still quite spaguetti (it was spaguetti before too).

Reply:
Before this PR I mean

Reply:
This reduces arguments passed to isClosest function.
We can simply call one isClosest function. Creating a struct and then calling two different functions is unnecessary extra complexity,